File: solutions/go/interest-is-interesting/1/interest_is_interesting.go
package interest

// InterestRate returns the interest rate for the provided balance.
func InterestRate(balance float64) float32 {
    var rate float32 = 0.0
    if balance < 0 {
        rate = 3.213
    } else if balance < 1000  {
        rate = 0.5
    } else if balance < 5000 {
        rate = 1.621
    } else {
        rate = 2.475
    }
    return rate
}

// Interest calculates the interest for the provided balance.
func Interest(balance float64) float64 {
    var rate     float64 = float64(InterestRate(balance))
    var interest float64 = balance * rate / 100
    return interest
}

// AnnualBalanceUpdate calculates the annual balance update, taking
// into account the interest rate.
func AnnualBalanceUpdate(balance float64) float64 {
    var interest float64 = Interest(balance)
    return balance + interest
}

// YearsBeforeDesiredBalance calculates the minimum number of years
// required to reach the desired balance.
func YearsBeforeDesiredBalance(balance, targetBalance float64) int {
    years := 0
    for balance < targetBalance {
        balance = AnnualBalanceUpdate(balance)
        years += 1
    }
    return years
}

File: solutions/cpp/interest-is-interesting/1/interest_is_interesting.cpp
// interest_rate returns the interest rate for the provided balance.
double interest_rate(double balance) {
    double rate{};
    if      ( balance <     0 ) { rate = 3.213; }
    else if ( balance < 1'000 ) { rate = 0.5;   }
    else if ( balance < 5'000 ) { rate = 1.621; }
    else                        { rate = 2.475; }
    return rate;
}

// yearly_interest calculates the yearly interest for the provided balance.
double yearly_interest(double balance) {
    double rate = interest_rate(balance);
    return balance * rate / 100;
}

// annual_balance_update calculates the annual balance update, taking into
// account the interest rate.
double annual_balance_update(double balance) {
    double interest = yearly_interest(balance);
    return balance + interest;
}

// years_until_desired_balance calculates the minimum number of years required
// to reach the desired balance.
int years_until_desired_balance(double balance, double target_balance) {
    double inc_balance = balance;
    int years{};
    while ( inc_balance < target_balance ) {
        years++;
        inc_balance = annual_balance_update(inc_balance);
    }
    return years;
}
